af/shovel/centrifugation.py

```diff
@@ -326,163 +329,491 @@ def __flush_stubborn(self):
-            datum_db = list(c)[0][0]
-            if datum_db != datum_count:
-                raise RuntimeError('Unexpected DB mismatch in measurement count', bucket, datum_count, datum_db)
-            if len(files) != 0:
-                c.execute('SELECT MIN(code_ver), MAX(code_ver) FROM autoclaved WHERE bucket_date = %s', (bucket,))
-                vermin, vermax = list(c)[0]
-            else:
-                vermin, vermax = CODE_VER, CODE_VER # as DB will return NULL in this case
-            if vermin != vermax:
-                raise RuntimeError('Unexpected DB mismatch in code versions', bucket, vermin, vermax)
-            return vermin
-        elif len(files_db) == 0: # and len(files) > 0 as sets are not equal :)
-            return None
+        c.execute('SELECT filename, file_sha1, code_ver FROM autoclaved WHERE bucket_date = %s', (bucket,))
+        for filename, file_sha1, code_ver in c:
+            file_sha1 = str(file_sha1) # file_sha1 is returned as buffer()
+            if filename not in files:
+                raise RuntimeError('Unknown filename in autoclaved PG table present in autoclaved/index.json.gz', filename)
+            elif files[filename] != file_sha1: # autoclaved file changed, e.g. PII cleanup
+                reingest.add(filename)
+            elif code_ver < CODE_VER: # including CODE_VER_REPROCESS
+                reprocess.add(filename)
+            elif code_ver == CODE_VER:
+                pass # no-op
+            else: # including code_ver > CODE_VER
+                raise RuntimeError('Bad autoclaved row', filename, file_sha1, code_ver)
+            files.pop(filename)
+    ingest = files.viewkeys() # files that are not in the database at all
+    return ingest, reingest, reprocess
+
+def create_tmp_table(pgconn, table, definition):
+    # XXX: never ever use user-supplied `new_table` and `tpl_table` :-)
+    if FLAG_TRUE_TEMP:
+        create_table = 'CREATE TEMPORARY TABLE {table} ({definition}) ON COMMIT DROP'
+    else:
+        create_table = 'CREATE TABLE {table} ({definition})'
+    with pgconn.cursor() as c:
+        c.execute(create_table.format(table=table, definition=definition))
+
+def copy_meta_from_index(pgconn, ingest, reingest, autoclaved_index, bucket):
+    # Writing directly to `report` table is impossible as part of `report`
+    # table is parsed from data files, same is true for `measurement` table.
+    # The reason for `autoclaved` is cross-reference generation, it's possible
+    # to avoid it, but the price is useless extra complexity. Filling these
+    # three tables one-by-one is also possible, but it makes re-ingestion
+    # preserving `*_no` sequence values and available range a more complex task
+    # as sequence is non-transactional.
+    if not DUPLICATE_REPORTS:
+        # FIXME: eventually this assertion should be removed and DUPLICATE_REPORTS
```

Comment:
I am not sure I agree with this. There are cases in which duplicate reports can also span multiple buckets (for example a user did an upload of a measurement they already submitted days after).

That said we should probably be detecting duplicates much earlier (maybe in the collector itself) and reject them there.